src/subsplit_dag.hpp

```diff
@@ -70,7 +71,58 @@ class SubsplitDAG {
   // Iterate over the node ids corresponding to rootsplits.
   void IterateOverRootsplitIds(const std::function<void(size_t)> &f) const;
 
-  // #288: consider renaming these.
+  // Apply an Action via a depth first traversal. Do not visit leaf nodes.
+  // Applied to a given node, we:
+  // * Apply BeforeNode
+  // * For each of the clades of the node, we:
+  //     * Apply BeforeNodeClade
+  //     * For each edge descending from that clade, we:
+  //         * Recur into the child node of the clade if it is not a leaf
+  //         * Apply VisitEdge to the edge
+  //     * Apply AfterNodeClade
+  // * Apply AfterNode
+  template <typename TraversalActionT>
+  void DepthFirstWithAction(const TraversalActionT &action) const {
+    std::unordered_set<size_t> visited_nodes;
+    for (const auto &rootsplit : rootsplits_) {
+      DepthFirstWithActionForNode(action, subsplit_to_id_.at(rootsplit + ~rootsplit),
+                                  visited_nodes);
+    }
+  };
+
+  // The portion of the traversal that is below a given node.
+  template <typename TraversalActionT>
+  void DepthFirstWithActionForNode(const TraversalActionT &action, size_t node_id,
+                                   std::unordered_set<size_t> &visited_nodes) const {
+    action.BeforeNode(node_id);
```

Comment:
I just have a stupid question here:

We could call `BeforeNode` and `AfterNode` because we know that TraversalActionT passed into this function is of type `SubsplitDAGTraversalAction`, is that correct?

Is there any way to ensure that the passed type `TraversalActionT` always has these functions? I guess I'm not quite sure why generic needs to be used? Why not just use `SubsplitDAGTraversalAction`?

Reply:
We have to use a template here because `SubsplitDAGTraversalAction` isn't a class, it's a template. This is necessary because the compiler needs to understand what's going into the lambdas in order to be able to build an object.

We _could_ make it a virtual class and go that route. There are pluses and minuses.

@ognian- said in chat

> Yes, when we remove runtime typing we need to template all usage sites. The example with std::function can be used without templating, if we remove the bundled state and loose it's lifetime safety. But that's only because std::function uses virtual dispatch under the hood and this causes larger object size and dynamic allocation, leading to call overhead and cache misses. This shouldn't be allowed in the DAG traversal hot path.
> The same basic rules applies to all compile-time polymorphism, if CRTP was used it will still require templating the usage sites. C++20 has a lighter syntax for this, where a function parameter can be just declared "auto" and the compiler will turn the function into a template (trough the new concepts mechanism).

Regarding 

> Is there any way to ensure that the passed type TraversalActionT always has these functions?

Yes, it's called "concepts" and will be available to us in C++20.